### PR TITLE
Added Limit/Sort pushdown for SQLite and conversion from SQLite numerical value to PostgreSQL TimeStamp

### PR DESCRIPTION
**Pull request by pruiz**
Added functionality to allow pushdown for Limit (Limit/Offset) and Order By.

NULLS FIRST/LAST also allowed (pushdown) if SQLite version has functionality implemented (>3.30).

Also, added conversion when adding foreign SQLite table that has numerical values to PostgreSQL timestamp, so now we can store TimeStamp in SQLite as TEXT or NUMERICAL and both will be working as expected with PostgreSQL.

PS: Tests Working with PostgreSQL 12 and SQLite 3.30

### CODE REVIEW

**Comment by hrkuma** [comment]
Thank you for implementing and sharing such good features!
We'd like to review and want to apply.
However it may take some time, please be patient.

File: sqlite_fdw.c [hunk 1]
```diff
@@ -365,7 +385,144 @@ sqliteGetForeignRelSize(PlannerInfo *root, RelOptInfo *baserel, Oid foreigntable
...
+	rows = startup_cost = total_cost = 10;
+
+	useful_pathkeys_list = get_useful_pathkeys_for_relation(root, rel);
+
+	/* Create one path for each set of pathkeys we found above. */
+	foreach(lc, useful_pathkeys_list)
+	{
+		List	   *useful_pathkeys = lfirst(lc);
+
+		if (rel->reloptkind == RELOPT_BASEREL ||
```
Comment (hrkuma) [inline sqlite_fdw.c:469]: Please use IS_SIMPLE_REL()

**Comment by pruiz** [comment]
> Thank you for implementing and sharing such good features!
> We'd like to review and want to apply.
> However it may take some time, please be patient.

Sure, np :)

File: sqlite_fdw.c [hunk 2]
```diff
@@ -376,11 +533,22 @@ sqliteGetForeignPaths(PlannerInfo *root, RelOptInfo *baserel, Oid foreigntableid
 {
 	Cost		startup_cost = 10;
 	Cost		total_cost = baserel->rows + startup_cost;
+	List	   *fdw_private = NIL;
 
 	elog(DEBUG1, "sqlite_fdw : %s", __func__);
 	/* Estimate costs */
 	total_cost = baserel->rows;
 
+	/* XXX: We add fdw_private with has_limit: true if these three conditions are true because we need to be able to pushdown
```
Comment (hrkuma) [inline sqlite_fdw.c:542]: Please delete "XXX:", we'd like to follow standard comment style.

File: deparse.c [hunk 3]
```diff
@@ -2467,13 +2456,65 @@ appendOrderByClause(List *pathkeys, deparse_expr_cxt *context)
 		else
 			appendStringInfoString(buf, " DESC");
 
-		if (pathkey->pk_nulls_first)
-			elog(ERROR, "NULLS FIRST not supported");
+		// XXX: In SQLITE3 Release v3.30.0 (2019-10-04) NULLS FIRST/LAST is supported, but not in prior versions
+		// More info: https://www.sqlite.org/changes.html https://www.sqlite.org/lang_select.html#orderby
+		int sqliteVersion = sqlite3_libversion_number();
```
Comment (hrkuma) [inline deparse.c:2461]: Please place the variable declarations in the beginning of the scope.

File: Makefile.extra [hunk 4]
```diff
@@ -0,0 +1,49 @@
+######################################################################-------------------------------------------------------------------------
```
Comment (hrkuma) [inline Makefile.extra:1]: Please remove three files (Makefile.extra, run_base_tests.sh and run_extra_tests.sh)

You have added the following new files.
- Makefile.extra (exactly same as existing Makefile)
- run_base_tests.sh (almost same as existing test.sh except for sed REGRESS replacement)
- run_extra_tests.sh (almost same as existing test_extra.sh except for final line and exporting USE_PGXS)
However these are almost same with existing files, so I don't think these files are required.
If the problem is just that test.sh cannot be executed after text_extra.sh, I think we should modify test.sh adding sed REGRESS replacement.

**Comment by hrkuma** [comment]
Thank you for fixing.

Could you add the description about the column_type in README.md ?

**Review by hrkuma — APPROVED** [review]
Thank you for update.
I'll merge soon.

File: sql/aggregate.sql [hunk 5]
```diff
@@ -1,3 +1,9 @@
+--SET log_min_messages  TO DEBUG1;
+--SET client_min_messages  TO DEBUG1;
+CREATE EXTENSION sqlite_fdw;
```
Comment (mlt) [inline sql/aggregate.sql:3]: Why is it here? Extension is already loaded in `sqlite_fdw.sql`

**Pull request merged**